### DISCUSSION TITLE
Hit a brick wall, how can I diagnose my issue

### DISCUSSION THREAD

**Discussion by dipunm** [micropython/micropython#10712] [RP2040 / Pico]
Hey all,

I am creating a custom firmware for the badger2040 which has the RP2040 embedded.

I am trying to get some C libraries pre-installed onto the device, and unfortunately, the device crashes (or hangs) when I try to use the library through the REPL connected over USB.

Here are a bunch of observations I have made:

1. Using the _thread libraries, if I use the C library on a separate thread, the REPL remains responsive but the thread never terminates (I think it crashes).
2. There are NO error messages, so debugging is really hard right now.
3. I am able to modify the source code, and I noticed (after putting a bunch of RETURN statements everywhere) that the crash happens on a variable assignment operation (assigning a property of newly allocated ram for an object) -- It makes me think there might be an out of memory or bad memory assignment issue, but --- no errors.

For context, here is a link to my github page which has Github Actions doing the build for me: https://github.com/dipunm/bitcoin-starter-kit

1. Is it normal to not get any errors from C or C++ code compiled into the firmware? 
2. Is there a tool I can use to get error messages?
3. Does this description give any of you with experience any ideas of what could be happening?

Thanks all in advance.

**Comment by jimmo** [comment]
hi @dipunm It would be helpful if you can provide more information about what you're doing. (I can't just go browsing around your repo)

One big thing that might be relevant -- Are you aware of how root pointers work in MicroPython, and does your code handle them as necessary?

> Is it normal to not get any errors from C or C++ code compiled into the firmware?

What do you mean by this? What are you expecting? If you call mp_printf() in your C code then it will be printed to the console.

You can also quite easily set up SWD to debug your C code via GDB on the RP2040.

**Comment by dipunm** [comment]
@jimmo thanks for the reply. I am very new to this, I am kind of building my knowledge as I go.

> provide more information about what you're doing. 

I am creating a custom firmware. I am trying to get a library designed for micropython into this firmware called `sec256k1-embedded`. This has AFAIK been used for other micropython models. I want to use a library called `embit` which uses this library as a dependency, but for now, I just need this library to compile and for me to be able to call its functions and get a result.

> Are you aware of how root pointers work in MicroPython

Nope, never heard of that, I try to find some resources to learn what that is. Mind you, the code I am writing is in python, the code that is not working is in C, but it is not my code. I am really hoping that there's something I'm doing wrong with the configuration or the build that is causing this and this is why I am debugging (as best I can).

> What do you mean by this? What are you expecting? If you call mp_printf() in your C code then it will be printed to the console.

I _think_ I did try `mp_printf()` but I will try again, I was trying to find a way to print useful comments but from memory I think I may have been trying to call `printf` instead. I gave up on that since it didn't print anything into the THONNY terminal.

> You can also quite easily set up SWD to debug your C code via GDB on the RP2040.

Could you please expand those abbreviations? Is SWD a product that I need to buy?

If it requires me to build code locally and run it, I may be a little out of luck since I had trouble doing this and I was advised to use Github Actions in a different Github Issue.
